napari/layers/shapes/_shapes_mouse_bindings.py

```diff
@@ -518,6 +524,9 @@ def vertex_insert(layer: Shapes, event: MouseEvent) -> None:
     with layer.events.set_data.blocker():
         layer._data_view.edit(index, vertices, new_type=new_type)
         layer._selected_box = layer.interaction_box(layer.selected_data)
+    layer.events.data(
+        value=layer.data, action_type={"action": "change", "index": [index]}
```

Comment:
I'm confused — is index the index of the shape within the shape list or the index of the vertex within the shape? Either way, it might be interesting to have *both*?

Reply:
I also have this nagging feeling that events should only be emitted from the Shapes layer itself and that this function should use APIs of the layer to edit the shape, and the event would be emitted as a matter of course...

Reply:
It is solely the index of the shape within the shape list. If we would add also vertex it would deviate from the event emission from points data, which would be fine. However, would you emit something like `vertex_indices` as well when moving the whole shape or solely when removing / inserting a vertex?

Reply:
@jni After discussion on zulip with @DragaDoncila we came up with multiple potential solutions. What is possible is to have 
```
layer.events.data(
            value=layer.data,
            action="change",
            data_indices=list(layer.selected_data),
        )
```

we could then add also `vertex_indices` which would be a `list` of lists to account for multiple shapes being moved. This could be iterated over by means of `enumeration`. What do you think?

We also thought about a `dict` with data_indices as keys and vertex_indices as values, but this is less generalizable between adding / removing a shape and when just adding / removing vertices of a shape. With the former solution vertex_indices would be optional.

Reply:
> I also have this nagging feeling that events should only be emitted from the Shapes layer itself and that this function should use APIs of the layer to edit the shape

Agree we should do this if possible. We had similar concerns with the `points` event and I remember doing some refactoring there to avoid emitting in the `mouse_bindings`

Reply:
However, we may not be certain yet how that would interface with the point on the roadmap regarding which parts of labels / shapes can be unified.

Reply:
I guess the question is should it be part of this PR which aims to change what is emitted.

Reply:
I think since the event was always being emitted here, we should leave it here in this PR and follow up with a refactoring PR for the `shapes` emitting. I did a quick search and it looks like it's only for the `shapes` layer that we still emit in the `mouse_bindings` so hopefully it wouldn't be too onerous a refactor. But yes, out of scope for this PR imo.

Reply:
Stating this as resolved as `vertex_indices` is now also emitted. I will open an issue for the refactor.